### PR TITLE
Update application environments

### PR DESCRIPTION
**Pull request by drummond-dds**
This PR allows editing application environments on the application settings page.

The AC for this PT task describes editing environment roles but this PR is just for editing the environment. We need to clarify the AC.

* [Pivotal Tracker](https://www.pivotaltracker.com/story/show/165067811)

### CODE REVIEW

**Comment by drummond-dds** [comment]
@dandds PR has been modified to be one form per environment rather than one form for the entire application. This should solve the nested form issue when updating an environments members.

File: templates/portfolios/applications/settings.html [hunk 1]
```diff
@@ -77,18 +77,25 @@ <h2 class="block-list__title">{{ 'portfolios.applications.environments_heading'
     </div>
   </form>
 
+  <div id="application-environments">
+    {% if g.matchesPath("application-environments") %}
+        {% include "fragments/flash.html" %}
```
Comment (dandds) [inline templates/portfolios/applications/settings.html]: The flash messages always display within the panel (unless you have a design that's an exception).

File: templates/fragments/applications/environments.html [hunk 2]
```diff
@@ -74,15 +76,16 @@
...
-                  </div>
+                  {% set edit_form = environment_info['edit_form'] %}
+                  <form action="{{ url_for('applications.update_environment', environment_id=environment_info['id']) }}" method="post">
+                    {{ edit_form.csrf_token }}
+                    {{ TextInput(edit_form.name) }}
+                    {{
+                      SaveButton(
+                        text=("portfolios.applications.update_button_text" | translate)
+                      )
+                    }}
```
Comment (dandds) [inline templates/fragments/applications/environments.html]: We should get designs for the placement of this button. Not necessary for this PR, but put it in the icebox please.

File: atst/domain/environments.py [hunk 3]
```diff
@@ -49,6 +49,14 @@ def for_user(cls, user, application):
             .all()
         )
 
+    @classmethod
+    def update(cls, environment, name=None):
+        if name is not None:
+            environment.name = name
+
+        db.session.add(environment)
+        db.session.commit()
```
[the remark this answers is not in the document]
Comment (drummond-dds) [reply atst/domain/environments.py]: Fixed in 9cbef13b

File: atst/routes/applications/settings.py [hunk 4]
```diff
@@ -62,6 +67,29 @@ def settings(application_id):
     )
 
 
+@applications_bp.route("/environments/<environment_id>", methods=["POST"])
+@user_can(Permissions.EDIT_APPLICATION, message="edit application environments")
```
[the remark this answers is not in the document]
Comment (drummond-dds) [reply atst/routes/applications/settings.py]: Fixed in d8265582

File: atst/routes/applications/settings.py [hunk 5]
```diff
@@ -62,6 +67,29 @@ def settings(application_id):
     )
 
 
+@applications_bp.route("/environments/<environment_id>", methods=["POST"])
```
[the remark this answers is not in the document]
Comment (drummond-dds) [reply atst/routes/applications/settings.py]: Fixed in 784b4188330f5bbfe1bc9785bc6c198cf498ea7c

File: templates/fragments/applications/environments.html (hunk 2 above)
Comment (drummond-dds) [reply templates/fragments/applications/environments.html]: Created https://www.pivotaltracker.com/story/show/165680581

File: templates/portfolios/applications/settings.html (hunk 1 above)
Comment (drummond-dds) [reply templates/portfolios/applications/settings.html]: Fixed in aa0e4899e8e4f8328ceba3fe7558c1557f0f3641

File: js/components/toggler.js [hunk 6]
```diff
@@ -1,6 +1,15 @@
+import FormMixin from '../mixins/form'
+import textinput from './text_input'
+
 export default {
   name: 'toggler',
 
+  mixins: [FormMixin],
```
[the remark this answers is not in the document]
Comment (drummond-dds) [reply js/components/toggler.js]: Yeah we need this for the save button toggle. It does seem weird but I'm not sure how we can alternatively do this without that mixin

File: atst/routes/applications/settings.py [hunk 7]
```diff
@@ -62,6 +67,29 @@ def settings(application_id):
...
+@user_can(Permissions.EDIT_APPLICATION, message="edit application environments")
+def update_environment(environment_id):
+    environment = Environments.get(environment_id)
+    application = environment.application
+
+    form = EditEnvironmentForm(obj=environment, formdata=http_request.form)
```
[the remark this answers is not in the document]
Comment (drummond-dds) [reply atst/routes/applications/settings.py]: Nope! Removed in 1d36ae76244cdbec12396dad01472b4686be4fe8

**Pull request merged**